File: Toolbox/Toolbox.py

from abc import ABC, abstractmethod

class ToolIf(ABC):
    @abstractmethod
    def get_cost(self):
        pass
    
    @abstractmethod
    def display(self):
        pass

class Hammer(ToolIf):
    def __init__(self, name, cost):
        self.name = name
        self.cost = cost
    
    def get_cost(self):
        return self.cost
    
    def display(self):
        print(f"{self.name} is a hammer and costs ${self.cost:.2f}.")

class Screwdriver(ToolIf):
    def __init__(self, name, cost):
        self.name = name
        self.cost = cost
    
    def get_cost(self):
        return self.cost
    
    def display(self):
        print(f"{self.name} is a screwdriver and costs ${self.cost:.2f}.")

def create_tool():
    tool = None
    select = 0
    name = ""
    cost = 0.0
    
    while select < 1 or select > 2:
        print()
        print("Enter 1 for a hammer, or")
        print("Enter 2 for a screwdriver.")
        select = int(input("Selection: "))
        
        if select == 1:
            name = input("What is the hammer's name? ")
            cost = float(input("How much does this hammer cost? $"))
            tool = Hammer(name, cost)
        elif select == 2:
            name = input("What is the screwdriver's name? ")
            cost = float(input("How much does this screwdriver cost? $"))
            tool = Screwdriver(name, cost)
        else:
            print("Invalid selection, try again!")
    
    return tool

class Toolbox:
    def __init__(self, name):
        self.name = name
        self.total_cost = 0.0
        self.items = []
        self.num_items = 0
    
    def add_tool(self, tool):
        if self.num_items < 4:
            self.items.append(tool)
            self.num_items += 1
            self.total_cost += tool.get_cost()
            return True
        return False
    
    def display(self):
        print()
        print(f"{self.name} is a toolbox with the following items:")
        for item in self.items:
            item.display()
        print(f"The total cost of these items is ${self.total_cost:.2f}\n")

if __name__ == "__main__":
    NUM_TOOLS = 2
    toolbox = Toolbox("Data Processing Support")
    tools = []

    for _ in range(NUM_TOOLS):
        tool = create_tool()
        ret = toolbox.add_tool(tool)
        if not ret:
            print("Unable to add the tool!")
        else:
            tools.append(tool)

    toolbox.display()

    for tool in tools:
        tool = None

File: Toolbox/Toolbox.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Toolbox {
    private String name;
    private double totalCost;
    private List<ToolIf> items;
    private int numItems;

    public Toolbox(String name) {
        this.name = name;
        this.totalCost = 0.0;
        this.items = new ArrayList<>();
        this.numItems = 0;
    }

    public boolean addTool(ToolIf tool) {
        if (numItems < 4) {
            items.add(tool);
            numItems++;
            totalCost += tool.getCost();
            return true;
        }
        return false;
    }

    public void display() {
        System.out.println();
        System.out.println(name + " is a toolbox with the following items:");
        for (ToolIf item : items) {
            item.display();
        }
        System.out.printf("The total cost of these items is $%.2f%n%n", totalCost);
    }

    public static void main(String[] args) {
        Toolbox toolbox = new Toolbox("Data Processing Support");
        ToolIf[] tools = new ToolIf[2];
        Scanner scanner = new Scanner(System.in);

        for (int i = 0; i < tools.length; i++) {
            ToolIf tool = createTool(scanner);
            boolean ret = toolbox.addTool(tool);
            if (!ret) {
                System.out.println("Unable to add the tool!");
            }
            tools[i] = tool;
        }

        toolbox.display();

        for (ToolIf tool : tools) {
            tool = null;
        }
    }

    private static ToolIf createTool(Scanner scanner) {
        ToolIf tool = null;
        int select;
        String name;
        double cost;

        do {
            System.out.println();
            System.out.println("Enter 1 for a hammer, or");
            System.out.println("Enter 2 for a screwdriver.");
            System.out.print("Selection: ");
            select = scanner.nextInt();
            scanner.nextLine();

            switch (select) {
                case 1:
                    System.out.print("What is the hammer's name? ");
                    name = scanner.nextLine();
                    System.out.print("How much does this hammer cost? $");
                    cost = scanner.nextDouble();
                    scanner.nextLine();
                    tool = new Hammer(name, cost);
                    break;
                case 2:
                    System.out.print("What is the screwdriver's name? ");
                    name = scanner.nextLine();
                    System.out.print("How much does this screwdriver cost? $");
                    cost = scanner.nextDouble();
                    scanner.nextLine();
                    tool = new Screwdriver(name, cost);
                    break;
                default:
                    System.out.println("Invalid selection, try again!");
            }
        } while (select < 1 || select > 2); // Repeat for an invalid selection

        return tool;
    }
}

interface ToolIf {
    double getCost();
    void display();
}

class Hammer implements ToolIf {
    private String name;
    private double cost;

    public Hammer(String name, double cost) {
        this.name = name;
        this.cost = cost;
    }

    @Override
    public double getCost() {
        return cost;
    }

    @Override
    public void display() {
        System.out.printf("%s is a hammer and costs $%.2f.%n", name, cost);
    }
}

class Screwdriver implements ToolIf {
    private String name;
    private double cost;

    public Screwdriver(String name, double cost) {
        this.name = name;
        this.cost = cost;
    }

    @Override
    public double getCost() {
        return cost;
    }

    @Override
    public void display() {
        System.out.printf("%s is a screwdriver and costs $%.2f.%n", name, cost);
    }
}
